scala/scala.bzl

```diff
@@ -494,6 +511,57 @@ def _scala_test_impl(ctx):
     _write_test_launcher(ctx, rjars)
     return _scala_binary_common(ctx, cjars, rjars)
 
+def _prep_grep_pattern_from(patterns):
+    combined_pattern = ""
+    for pattern in patterns:
+      combined_pattern += " -e {}\.class".format(pattern)
+    return combined_pattern
+
+def _discover_classes(ctx, patterns, archive):
+    discovered_classes = ctx.new_file(ctx.label.name + "_discovered_classes.txt")
+    ctx.action(
+      #We need _jdk to even run _jar. Depending on _jar is not enough with sandbox
+      inputs= [archive] + ctx.files._jar + ctx.files._jdk,
+      outputs=[discovered_classes],
+      progress_message="Discovering classes with patterns of %s" % patterns,
+      #TODO consider with Damien/Ulf/Oscar the implications of switching from grep to scala code
+      #Pro-> logic will be cohesive (currently the scala code assumes stuff from the grep)
+      #Con-> IIRC Ulf warned me about performance implications of these traversals
+      command="{jar} -tf {archive} | grep {combined_patterns} > {out}".format(archive=archive.path, combined_patterns=_prep_grep_pattern_from(patterns), out=discovered_classes.path, 
+      jar=ctx.file._jar.path))
```

Comment:
+1, since merely calling it to execute it here and I am targeting removing single_file=True in other PR.